module_utils/foreman_helper.py

```diff
@@ -51,22 +53,189 @@ def filter_module_params(self):
 
+
+class ForemanAnsibleModule(ForemanBaseAnsibleModule):
+
+    def check_requirements(self):
+        if not HAS_NAILGUN:
+            self.fail_json(msg='The nailgun Python module is required',
+                           exception=NAILGUN_IMP_ERR)
+
     def connect(self, ping=True):
         entity_mixins.DEFAULT_SERVER_CONFIG = ServerConfig(
             url=self._foremanapi_server_url,
             auth=(self._foremanapi_username, self._foremanapi_password),
             verify=self._foremanapi_verify_ssl,
         )
 
+    def ping(self):
+        try:
+            return Ping().search_json()
+        except Exception as e:
+            self.fail_json(msg="Failed to connect to Foreman server: %s " % e)
+
+
+class ForemanApypieAnsibleModule(ForemanBaseAnsibleModule):
+
+    def check_requirements(self):
+        if not HAS_APYPIE:
+            self.fail_json(msg='The apypie Python module is required',
+                           exception=APYPIE_IMP_ERR)
+
+    def connect(self, ping=True):
+        try:
+            self.foremanapi = apypie.Api(uri=self._foremanapi_server_url,
+                                         username=self._foremanapi_username,
+                                         password=self._foremanapi_password,
+                                         api_version=2,
+                                         verify_ssl=self._foremanapi_verify_ssl)
+        except Exception as e:
+            self.fail_json(msg="Failed to connect to Foreman server: %s " % e)
+
         if ping:
             self.ping()
 
     def ping(self):
         try:
-            return Ping().search_json()
+            return self.foremanapi.resource('home').call('status')
         except Exception as e:
             self.fail_json(msg="Failed to connect to Foreman server: %s " % e)
 
+    def show_resource(self, resource, resource_id):
+        return self.foremanapi.resource(resource).call('show', {'id': resource_id})
+
+    def find_resource(self, resource, search, failsafe=False, thin=False):
+        response = self.foremanapi.resource(resource).call('index', {'search': search, 'thin': thin})
+        results = response['results']
+        if len(results) == 1:
+            result = results[0]
+        elif failsafe:
+            result = None
+        else:
+            self.fail_json(msg="No data found for %s" % search)
+        if result:
+            if thin:
+                result = {'id': result['id']}
+            else:
+                result = self.show_resource(resource, result['id'])
+        return result
+
+    def find_resource_by_name(self, resource, name, failsafe=False, thin=False):
+        search = 'name="{}"'.format(name)
+        return self.find_resource(resource, search, failsafe, thin)
+
+    def find_resources(self, resource, search_list, failsafe=False, thin=False):
+        return list(map(lambda search_item: self.find_resource_by_name(resource, search_item, failsafe=failsafe, thin=thin), search_list))
```

Comment:
```suggestion
        return [self.find_resource_by_name(resource, search_item, failsafe=failsafe, thin=thin) for search_item in search_list]
```

I also wonder if you can do some smart searching like `search = ' or '.join('({})'.format(item) for item in search_list)` and only do one search.

Reply:
The other helper functions don't handle multiple results well, so this is why it's not been done like this yet. But for the future this is certainly a good idea.